### PR TITLE
feat: [icx-proxy] Add a list of canister ID aliases

### PR DESCRIPTION
**Pull request by ericswanson-dfinity**
Adds a `--dns-alias identity.ic0.app:aaaaa-aa` command line argument that maps domain names to canister IDs.

Implements https://github.com/dfinity/agent-rs/issues/149

### CODE REVIEW

File: icx-proxy/src/main.rs [hunk 1]
```diff
@@ -69,13 +71,27 @@ pub(crate) struct Opts {
...
+    dns_aliases: &DnsAliases,
+) -> Option<Principal> {
     let url = Uri::from_str(hostname).ok()?;
 
     // Check if it's localhost or ic0.
-    match url.host()?.split('.').collect::<Vec<&str>>().as_slice() {
+    let host_parts = url.host()?.split('.').collect::<Vec<&str>>();
+    let host_parts = host_parts.as_slice();
+
+    if let Some(principal) = dns_aliases.resolve_canister_id_from_host_parts(host_parts) {
```
Comment (hansl) [inline icx-proxy/src/main.rs:91]:
```suggestion
    if let Some(principal) = dns_aliases.resolve_canister_id_from_host_parts(&host_parts) {
```

File: icx-proxy/src/dns_aliases.rs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    // dns_suffix_slice2: &[&str],
+    principal: Principal,
+}
+
+#[derive(Clone, Debug)]
+pub struct DnsAliases {
+    dns_aliases: Vec<DnsAlias>,
+}
+
+impl DnsAliases {
```
Comment (hansl) [inline icx-proxy/src/dns_aliases.rs:27]: Add some comments here and for the pub methods of this struct.

File: icx-proxy/src/config/canister_dns_config.rs [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+impl CanisterDnsConfig {
+    /// Create a CanisterDnsConfig instance from command-line configuration.
+    /// dns_aliases: 0 or more entries of the form of dns.alias:canister-id
+    pub fn new(dns_aliases: &[String]) -> anyhow::Result<CanisterDnsConfig> {
+        let dns_aliases = dns_aliases
+            .iter()
+            .map(|alias| DnsAlias::new(alias))
+            .collect::<anyhow::Result<Vec<DnsAlias>>>()?;
+        Ok(CanisterDnsConfig { dns_aliases })
```
Comment (hansl) [inline icx-proxy/src/config/canister_dns_config.rs:18]: I filed https://github.com/dfinity/agent-rs/issues/153 which is non blocking for this but would solve a gotcha when passing subdomains as aliases.

**Review by hansl — APPROVED** [review]
One non blocking comment, if you fix it here that's great please mark the bug as fixed too.

Thanks for the good work!

**Comment by nomeata** [comment]
BTW, did you spot that we were using this feature (off this branch) in the Demo today, when Christoph browsed to `https://identity.localhost`?

**Comment by paulyoung** [comment]
@kritzcreek, could you help me understand how you were able to get `name.localhost` to work?

I tried `icx-proxy --dns-alias name.localhost:rrkah-fqaaa-aaaaa-aaaaq-cai` but requests made to it via `curl` result in `Could not resolve host: name.localhost`

Did you just add an entry to `/etc/hosts`, or something else?

**Comment by kritzcreek** [comment]
@paulyoung IIRC I just followed these steps at the time: https://github.com/dfinity/internet-identity/blob/cb75f38160ee60cfb0ab986b5cbabfff00d21e5e/Demo.md#setup

I definitely didn't mess with the hosts file.

**Comment by paulyoung** [comment]
Can anyone else help me understand how this is intended to work?

I think I followed the same steps as above but I can’t see how anything is supposed to be servicing requests at `x.localhost` domains.

**Comment by hansl** [comment]
`curl` could work if you use the `Host:` header and use `localhost` instead of trying to resolve a subdomain (which doesn't work in standard tooling).

**Comment by paulyoung** [comment]
Thanks @hansl

**Pull request merged**